src/NuGet.Core/NuGet.Packaging/Rules/NoRefOrLibFolderInPackageRule.cs

```diff
@@ -0,0 +1,52 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NuGet.Common;
+
+namespace NuGet.Packaging.Rules
+{
+    class NoRefOrLibFolderInPackageRule : IPackageRule
+    {
+        public string MessageFormat { get; }
+
+        public NoRefOrLibFolderInPackageRule(string messageFormat)
+        {
+            MessageFormat = messageFormat;
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            if (!builder.GetFiles()
+                .Select(t => PathUtility.GetPathWithDirectorySeparator(t))
+                .Any(t => t.StartsWith(PackagingConstants.Folders.Lib + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)
+                || t.StartsWith(PackagingConstants.Folders.Ref + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)))
+            {
+                //if you can't find the ref and lib folder, then find the build folder
+                if (builder.GetFiles()
+                .Select(t => PathUtility.GetPathWithDirectorySeparator(t))
+                .Any(t => t.StartsWith(PackagingConstants.Folders.Build + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)))
+                {
+                    //if you can't find any folders other than native or any, raise an NU5127
+                    if (!builder.GetFiles()
+                    .Select(t => PathUtility.GetPathWithDirectorySeparator(t))
+                    .Where(t => t.StartsWith(PackagingConstants.Folders.Build + Path.DirectorySeparatorChar))
+                    .Any(t => !t.StartsWith(PackagingConstants.Folders.Build + Path.DirectorySeparatorChar + PackagingConstants.Folders.Native + Path.DirectorySeparatorChar
+                    , StringComparison.OrdinalIgnoreCase)
+                    && !t.StartsWith(PackagingConstants.Folders.Build + Path.DirectorySeparatorChar + PackagingConstants.Folders.Any + Path.DirectorySeparatorChar
```

Comment:
To my understanding, there is a case where a build folder can contain an `any/` directory in the 'build/' folder. We could change the warning to reference the tfm found in the build directory, Is there a case where the build directory will just have a .dll with no tfm referenced?

Reply:
well, normally the build directory will contain MSBuild props and targets files, rather than dlls, but the package author might include dlls as well, which they'll need to manage themselves since NuGet only looks for targets and props files. But yes, sometimes people have a `build/packageid.targets` file, so no subdirectory under `build/`.

It's a different validation scenario, but maybe we should also warn if a package contains:
```text
lib/netstandard2.0/packageid.dll
build/packageid.targets`
```
because the `lib/` directory makes it look like it's targeting `netstandard2.0`, but the `build/` directory looks like a catch-all.

Reply:
> It's a different validation scenario, but maybe we should also warn if a package contains:

I don't think that's a warning. 

lib/ref are first class asset group. If they exist you have to select one from there, otherwise the package is incompatible. 

build/*.targets in that case is ok, it doesn't cause any issues.

Reply:
my point is that if a package contains `lib/netstandard2.0/packageid.dll` and `build/packageid.targets`, it's not clear if the package author intended the build file to cause the package to be compatible with all TFMs, but have the lib used only for `netstandard2.0` compatible projects. Since the package will only be compatible with `netstandard2.0` they could put the targets file in `build/netstandard2.0/` to make their intent explicit. The warning would be to ensure that the package author understands how the package will behave.

Reply:
The trick is the majority of the packages are authored with build/* (I'll have exact numbers in an hour!). 
So adding a warning for something that doesn't cause problems and impacts the vast majority of customers is...well tricky :)